gobblin-service/src/main/java/org/apache/gobblin/service/modules/flow/FlowGraphPathFinder.java

```diff
@@ -0,0 +1,428 @@
+
+  /**
+   * A helper class used to maintain additional context associated with each {@link FlowEdge} during path
+   * computation while the edge is explored for its eligibility. The additional context includes the parent
+   * {@link FlowEdge}, the input {@link DatasetDescriptor} of this edge which is compatible with the parent's output
+   * {@link DatasetDescriptor}, and the corresponding output dataset
+   * descriptor of this edge. Whether or not a {@link FlowEdge} is added to the edge queue depends on whether
+   */
+  @Data
+  @AllArgsConstructor
+  @Getter
+  public static class FlowEdgeContext {
+    private FlowEdge edge;
+    private DatasetDescriptor inputDatasetDescriptor;
+    private DatasetDescriptor outputDatasetDescriptor;
+  }
+
+  /**
+   * A simple path finding algorithm based on Breadth-First Search. At every step the algorithm adds the adjacent {@link FlowEdge}s
+   * to a queue. The {@link FlowEdge}s whose output {@link DatasetDescriptor} matches the destDatasetDescriptor are
+   * added first to the queue. This ensures that dataset transformations are always performed closest to the source.
+   * @return a list of {@link FlowEdge}s starting at the srcNode and ending at the destNode.
+   */
+  public Dag<JobSpecWithExecutor> findPath() throws PathFinderException {
+    try {
+      //Initialization of auxiliary data structures used for path computation
+      this.pathMap = new HashMap<>();
+      this.flowEdgeToSpecExecutorMap = new HashMap<>();
+      this.flowEdgeToMergedConfig = new HashMap<>();
+
+      // Generate flow execution id for this compilation
+      this.flowExecutionId = System.currentTimeMillis();
+
+      //Path computation must be thread-safe to guarantee read consistency. In other words, we prevent concurrent read/write access to the
+      // flow graph.
+      // TODO: we can easily improve the performance by using a ReentrantReadWriteLock associated with the FlowGraph. This will
+      // allow multiple concurrent readers to not be blocked on each other, as long as there are no writers.
+      synchronized (this.flowGraph) {
+        //Base condition 1: Source Node or Dest Node is inactive; return null
+        if (!srcNode.isActive() || !destNode.isActive()) {
+          log.warn("Either source node {} or destination node {} is inactive; skipping path computation.", srcNodeId,
+              destNodeId);
+          return null;
+        }
+
+        //Base condition 2: Check if we are already at the target. If so, return an empty dag.
+        if ((srcNode.equals(destNode)) && srcDatasetDescriptor.isCompatibleWith(destDatasetDescriptor)) {
+          return new Dag<>(new ArrayList<>());
+        }
+
+        LinkedList<FlowEdgeContext> edgeQueue = new LinkedList<>();
+        edgeQueue.addAll(getNextEdges(srcNode, srcDatasetDescriptor, destDatasetDescriptor, flowConfig));
+        for (FlowEdgeContext flowEdgeContext : edgeQueue) {
+          this.pathMap.put(flowEdgeContext, flowEdgeContext);
```

Comment:
Instead of using a map to keep track of all links, can we just use a list or queue to keep track of the path we're interested?

Reply:
@zxcware I think this is fine because the map is used to recover the original path.

Reply:
The search is a depth first search, aggressively going deeper to find the first working path. We can push an candidate edge to a queue, popping it out if it can't match further. In the end, the queue describes all hops of a result path.